### ISSUE TITLE
Docker images with Java 8?

### ISSUE THREAD

**Issue by efenderbosch** [killbill/killbill-cloud#63]
Any plans on releasing Docker images that use Java 8? We're implementing our own payment plugin and it uses libraries that require Java 8. I've hacked Java 8 in to the existing Docker image, but it would be nice to have the option of picking Ubuntu 14 or 16 LTS or Alpine and Java 7 or 8.

I can probably help by forking and submitting a pull request.

**Comment by pierre** [comment]
> it would be nice to have the option of picking Ubuntu 14 or 16 LTS or Alpine and Java 7 or 8.

Switching to Ubuntu 16.04 LTS + Java 8 would make sense indeed. We could time releasing the new Docker images with the release of Kill Bill 0.18.0 (in a few weeks).

> I can probably help by forking and submitting a pull request.

👍  Thanks!

**Comment by efenderbosch** [comment]
I worked on 16.04 LTS and Java 8 as well as Alpine today. Alpine is more difficult due to it being so thin. I might have to put this on the back burner next week, but come back to it shortly.

**Comment by pierre** [comment]
> I worked on 16.04 LTS and Java 8 as well as Alpine today. Alpine is more difficult due to it being so thin.

The topic of Alpine support has come up on the mailing-list. My main concern is that it would ship a custom built JVM.

I think we will want to keep the Ubuntu images as the default, official, ones (even if they are quite large). No problem in helping the community maintain an Alpine Dockerfile though, as long as we find a way to use inheritance or build tricks to avoid duplicating lots of logic.

> I might have to put this on the back burner next week, but come back to it shortly.

No rush, the team is busy at [Money20/20](https://www.money2020.com/) anyways! 😸

**Comment by efenderbosch** [comment]
What version of tomcat should be used for Ubuntu 16?

**Comment by pierre** [comment]
It looks like v7 is still the version supported (`tomcat7` package, see https://help.ubuntu.com/lts/serverguide/tomcat.html)

**Comment by efenderbosch** [comment]
How important is `docker/templates/build/Dockerfile`?

I've been trying to fix the build that happens when this image is built by adding many `gem install ...` commands, but there are conflicting versions all over the place.

I think that `docker/templates/base/latest/Dockerfile` is ready and that's the important file. I'll probably submit a pull request shortly. Should I include the mess in `docker/templates/build/Dockerfile` or just leave it out?

**Comment by pierre** [comment]
> How important is `docker/templates/build/Dockerfile`?

Critical 😸  It powers our build environment (where we build the official binaries).

> I've been trying to fix the build that happens when this image is built by adding many `gem install ...` commands, but there are conflicting versions all over the place.

Can you go into details regarding these conflicts?

> I think that `docker/templates/base/latest/Dockerfile` is ready and that's the important file. I'll probably submit a pull request shortly. Should I include the mess in `docker/templates/build/Dockerfile` or just leave it out?

Better to have a clean PR. Keeping the build environment unchanged is fine - let's just make sure that there is no build issue because of inheritance of the `Dockerfile`.

**Comment by efenderbosch** [comment]
Current diff of that Dockerfile looks like this:

[diff.txt](https://github.com/killbill/killbill-cloud/files/571583/diff.txt)

And there are still errors in the build:

`
Bundler::GemspecError: Could not read gem at
/var/lib/jruby/lib/ruby/gems/shared/cache/rake-10.5.0.gem. It may be corrupted.
An error occurred while installing rake (10.5.0), and Bundler cannot continue.
Make sure that gem install rake -v '10.5.0' succeeds before bundling.
`

And it seems like every time I include another gem, that just causes the build to find others that are missing or have a version conflict.

**Comment by pierre** [comment]
Could you specify the version when installing bundler, etc (just use the versions used in the current image)?

```
# Install Bundler
RUN jruby -S gem install bundler -v XXX
# Install JBundler
RUN jruby -S gem install jbundler -v YYY
# Install a JavaScript runtime
RUN jruby -S gem install therubyrhino -v ZZZ
```

I guess it's the latest gems themselves causing these errors, not the upgrade of Ubuntu or JRuby.

**Comment by efenderbosch** [comment]
I could try that. What are those versions, though? Or how can I find that info?

edit: I'll pull from master, do that build and then pin bundler, jbundler and therubyrhino to the versions in that image.

**Comment by efenderbosch** [comment]
I think I figured it out. sudo behavior was changed between Ubuntu 14 and 16. Simply being in the sudo group is not sufficient to run commands via sudo. The user/group also has to be in `/etc/sudoers`. bundler was not able to install gems as root, which was causing all the build problems I was seeing. Adding this line to the killbill user section seems to have worked:

`echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers`

Ideally that should be locked down the to killbill user and just the commands necessary, but since this just a build image and isn't a distributed the above should be sufficient.
